catch/unit/warp/warp_common.hh

```diff
@@ -0,0 +1,141 @@
+  const auto idx = block_rank * warps_per_block + block.thread_rank() / warpSize;
+
+  return !(active_masks[idx] & (static_cast<uint64_t>(1) << warp.thread_rank()));
+}
+
+static inline std::mt19937& GetRandomGenerator() {
+  static std::mt19937 mt(11);
+  return mt;
+}
+
+template <typename T> static inline T GenerateRandomInteger(const T min, const T max) {
+  std::uniform_int_distribution<T> dist(min, max);
+  return dist(GetRandomGenerator());
+}
+
+inline uint64_t get_predicate_mask(unsigned int test_case, unsigned int warp_size) {
+  uint64_t predicate_mask = 0;
+  switch (test_case) {
+    case 0:  // no thread
+      predicate_mask = 0;
+      break;
+    case 1:  // 1st thread
+      predicate_mask = 1;
+      break;
+    case 2:  // last thread
+      predicate_mask = static_cast<uint64_t>(1) << (warp_size - 1);
+      break;
+    case 3:  // all threads
+      predicate_mask = 0xFFFFFFFFFFFFFFFF;
+      break;
+    default:  // random
+      static std::mt19937_64 mt(test_case);
+      std::uniform_int_distribution<uint64_t> dist(0, std::numeric_limits<uint64_t>::max());
+      predicate_mask = dist(mt);
+  }
+  return predicate_mask;
+}
+
+inline uint64_t get_active_predicate(uint64_t predicate, size_t partition_size) {
+  uint64_t active_predicate = predicate;
+  for (int i = partition_size; i < 64; i++) {
+    active_predicate &= ~(static_cast<uint64_t>(1) << i);
+  }
+  return active_predicate;
+}
+
+inline int generate_width(int warp_size) {
+  int exponent = 0;
+  while (warp_size >>= 1) {
+    ++exponent;
+  }
+
+  return GENERATE_COPY(map([](int e) { return 1 << e; }, range(1, exponent + 1)));
+}
+
+template <typename Derived, typename T> class WarpTest {
+ public:
+  WarpTest() : warp_size_{get_warp_size()} {}
+
+  void run() {
+    const auto blocks = GenerateBlockDimensionsForShuffle();
+    INFO("Grid dimensions: x " << blocks.x << ", y " << blocks.y << ", z " << blocks.z);
+    const auto threads = GenerateThreadDimensionsForShuffle();
+    INFO("Block dimensions: x " << threads.x << ", y " << threads.y << ", z " << threads.z);
+    grid_ = CPUGrid(blocks, threads);
+
+    const auto alloc_size = grid_.thread_count_ * sizeof(T);
+    LinearAllocGuard<T> arr_dev(LinearAllocs::hipMalloc, alloc_size);
+    LinearAllocGuard<T> arr(LinearAllocs::hipHostMalloc, alloc_size);
+    HIP_CHECK(hipMemset(arr_dev.ptr(), 0, alloc_size));
+
+    warps_in_block_ = (grid_.threads_in_block_count_ + warp_size_ - 1) / warp_size_;
+    const auto warps_in_grid = warps_in_block_ * grid_.block_count_;
+    LinearAllocGuard<uint64_t> active_masks_dev(LinearAllocs::hipMalloc,
+                                                warps_in_grid * sizeof(uint64_t));
+    active_masks_.resize(warps_in_grid);
+    std::generate(active_masks_.begin(), active_masks_.end(),
+                  [] { return GenerateRandomInteger(0ul, std::numeric_limits<uint64_t>().max()); });
```

Comment:
We should avoid using only random values as input because that hurts reproducibility.

Reply:
> We should avoid using only random values as input because that hurts reproducibility.

Agree that not only random patterns need to be checked.  Do you have particular patterns in mind though?

Reply:
@scchan @b-sumner Test has been expanded to have a version that uses random inputs and random active_masks, and version that uses predefined active masks and inputs (5 different patterns for active masks have been chosen and thread id has been used as input data (changed from warp id to include larger input values).